core/src/main/scala/com/gu/tableversions/core/InMemoryTableVersions.scala

```diff
@@ -17,95 +13,63 @@ import com.gu.tableversions.core.util.RichRef._
+    val applyUpdate: TableUpdates => Either[Exception, TableUpdates] = { currentTableUpdates =>
+      if (currentTableUpdates.contains(table)) {
+        val updated = currentTableUpdates + (table -> TableState(
+          currentVersion = update.header.id,
+          updates = currentTableUpdates(table).updates :+ update))
+        Right(updated)
+      } else
+        Left(unknownTableError(table))
     }
 
-  override def currentVersion(table: TableName): F[TableVersion] =
-    // Derive current version of a table by folding over the history of changes
-    for {
-      allTableUpdates <- allUpdates.get
-      tableUpdates <- allTableUpdates
-        .get(table)
-        .fold(F.raiseError[List[TableUpdate]](new Exception(s"Table '$table' not found")))(F.pure)
-      operations = tableUpdates.flatMap(_.operations)
-    } yield
-      if (isSnapShot(operations))
-        InMemoryTableVersions.latestSnapshotTableVersion(operations)
-      else
-        InMemoryTableVersions.applyPartitionUpdates(PartitionedTableVersion(Map.empty))(operations)
-
-  private def isSnapShot(operations: List[TableOperation]) = operations match {
-    case InitTable(_, isSnapshot) :: _ => isSnapshot
-    case _                             => throw new IllegalArgumentException("First operation should be InitTable")
+    allUpdates.modifyEither(applyUpdate).void
   }
 
-  override def commit(table: TableName, update: TableVersions.TableUpdate): F[TableVersions.CommitResult] = {
-
-    // Note: we're not checking invalid partition versions here as we suspect this problem will go
-    // away in a later iteration of this interface.
-
+  override def setCurrentVersion(table: TableName, id: CommitId): F[Unit] = {
     val applyUpdate: TableUpdates => Either[Exception, TableUpdates] = { currentTableUpdates =>
-      if (currentTableUpdates.contains(table)) {
-        val updated = currentTableUpdates + (table -> (currentTableUpdates(table) :+ update))
-        Right(updated)
-      } else
-        Left(new Exception(s"Unknown table '${table.fullyQualifiedName}'"))
+      currentTableUpdates.get(table) match {
+        case Some(currentTableState) =>
+          if (currentTableState.updates.exists(_.header.id == id)) {
+            val newTableState = currentTableState.copy(currentVersion = id)
+            val updated = currentTableUpdates + (table -> newTableState)
+            Right(updated)
+          } else
+            Left(unknownCommitId(id))
+
+        case None => Left(unknownTableError(table))
+      }
     }
 
-    allUpdates.modifyEither(applyUpdate).as(SuccessfulCommit)
+    allUpdates.modifyEither(applyUpdate).void
   }
 
+  override def tableState(table: TableName): F[TableState] =
+    for {
+      allTableUpdates <- allUpdates.get
+      tableState <- allTableUpdates
+        .get(table)
+        .fold(F.raiseError[TableState](unknownTableError(table)))(F.pure)
+    } yield tableState
```

Comment:
Not a request for change. Just a stylistic question on the back of that other discussion about `Ref` and exceptions.

Seeing we're in `F` here (unlike yesterday we were still in `Ref`) and the same law `F.delay[A](throw e) <-> F.raiseError(e)` should apply for `IO`, is the following easier to read, as in 
* no `Option.fold()` 
* only 1 explicit call to `F....` in lieu of 2, 
*  line shorter)

```
for {
      allTableUpdates <- allUpdates.get
      tableState <- F.delay(allTableUpdates(table)).adaptError {
        case _ => unknownTableError(table)
      }
    } yield tableState
```

Reply:
Actually, since I wrote that I have learnt that there is an `F.fromOption()`` method which would simplify this. In which case this becomes:

```scala
    for {
      allTableUpdates <- allUpdates.get
      tableState <- F.fromOption(allTableUpdates.get(table), unknownTableError(table))
    } yield tableState
```

Which is nice and simple?

Reply:
That's beautiful ;-)

Reply:
Can't tell if sarcasm or not... 😁

Reply:
It wasn't. `F.fromOption` is the better version here